Lib/tempfile.py

```diff
@@ -822,8 +822,20 @@ def resetperms(path):
 
                     try:
                         _os.unlink(path)
-                    # PermissionError is raised on FreeBSD for directories
-                    except (IsADirectoryError, PermissionError):
+                    except IsADirectoryError:
+                        cls._rmtree(path, ignore_errors=ignore_errors)
+                    except PermissionError:
+                        # The PermissionError handler was originally added for
+                        # FreeBSD in directories, but it seems that it is raised
+                        # on Windows too.
+                        # bpo-43153: Calling _rmtree again may
+                        # raise NotADirectoryError and mask the PermissionError.
+                        # So we must re-raise the current PermissionError if
+                        # path is not a directory.
+                        if _os.path.islink(path) or not _os.path.isdir(path):
+                            if ignore_errors:
+                                return
```

Comment:
Note that this check is inconsistent with `shutil.rmtree()`. `os.unlink()` deletes a mount point in Windows, but otherwise a mount point is handled as a directory,  i.e. `islink()` will be false and `isdir()` will be true. However, they're special cased by `shutil._rmtree_isdir()` and `shutil._rmtree_islink()`, and thus `shutil.rmtree()` will immediately fail with `OSError("Cannot call rmtree on a symbolic link")`, as would happen for a regular symbolic link, which is exactly what this check is supposed to prevent.

Reply:
To clarify, so are you suggesting we:
1. Modify shutil._rmtree_isdir with your suggested changes here https://github.com/python/cpython/pull/29940#discussion_r764529814.
2. Use shutil._rmtree_isdir for a more consistent check instead of our current method?

Reply:
I know it's a bad practice to rely on a private function from another module. But duplicating the code would be a maintenance burden, and it's not useful enough to expose in `os.path`. Maybe `_rmtree_isdir()` and `_rmtree_islink()` should be refactored into a common private module. I'm curious to know what @serhiy-storchaka thinks should be done about the consistency problem between these functions and `os.path.islink()` and `os.path.isdir()`.

Reply:
I'm perfectly fine with using private functions from another module as long as it's within the stdlib itself and not exposed somewhere :). Thank you for the clarification, I wanted to summarize your suggestions since I wasn't sure I understood you correctly.

Reply:
I think it should be okay to use shutil._rmtree_isdir() here because we're asking shutil if it's valid to use rmtree on a path, and then telling shutil to use shutil._rmtree(). So these two steps of logic stay completely internal to shutil even though used outside of it. And it's also consistent with current logic of using internal shutil._rmtree().

Reply:
`_rmtree_isdir()` and `_rmtree_islink()` was changed between versions, and this fix should be applied to multiple Python versions. So it is better to inline them here. If possible, we should use public API (add it if it does not exist).